### PR TITLE
Payments

### PR DESCRIPTION
**Pull request by jdungan**
Tox won't run on my workstation after the upgrade to python 2.7.11.  Tests and flake8 pass.  Hopefully it runs for you.

### CODE REVIEW

File: .env-dist [hunk 1]
```diff
@@ -9,3 +9,10 @@ GITHUB_CLIENT_SECRET=08c3da1421bb280e6fa5f61c05afd0c3128a2f9f
 SENDGRID_USERNAME="https://sendgrid.com/user/signup"
 SENDGRID_PASSWORD="https://sendgrid.com/user/signup"
 TEST_RUNNER="django_nose.NoseTestSuiteRunner"
+STRIPE_TEST_SECRET_KEY=""
+STRIPE_TEST_PUBLIC_KEY=""
+STRIPE_LIVE_SECRET_KEY=""
+STRIPE_LIVE_PUBLIC_KEY=""
```
Comment (groovecoder) [inline .env-dist]: Make just 'STRIPE_SECRET_KEY` and `STRIPE_PUBLIC_KEY`

File: api/views.py [hunk 2]
```diff
@@ -208,10 +218,13 @@ def post(self, request, format=None):
...
+                if item.transaction_status == "SUCCESS":
+                    new_payout.confirmation = item.payout_item_id
+                new_payout.save()
+                claim.status = "Paid"
+                claim.save()
+        else:
+            pass
 
-        return redirect('ClaimAPIView')
+        return redirect('/claim-status/' + str(claim.id))
```
Comment (groovecoder) [inline api/views.py]: Should use `reverse(`claim-status', pk=claim.id)`

File: auctions/models.py [hunk 3]
```diff
@@ -339,7 +333,8 @@ class Payment(models.Model):
         ('Stripe', 'Stripe'),
         ('PayPal', 'PayPal'),
     )
-    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False),
+    id = models.UUIDField(
+        primary_key=True, default=uuid.uuid4(), editable=False),
```
Comment (groovecoder) [inline auctions/models.py]: Should be able to use `default=uuid.uuid4`, or something like `default=generate_id` and specify a `def generate_id` function.

File: api/views.py [hunk 4]
```diff
@@ -181,38 +181,50 @@ class PayoutViewSet(APIView):
 
     def post(self, request, format=None):
         claim = Claim.objects.get(id=request.POST['claim'])
+
+        # leave if claim already paid
+        if claim.status == 'Paid':
+            return redirect('/claim-status/' + str(claim.id))
```
Comment (groovecoder) [inline api/views.py]: Should be `reverse('claim-status', pk=claim.id)`

**Comment by groovecoder** [comment]
Looks like the `claim-status` template is showing the "Your claim was approved. Would you like to get paid?" message to all viewers?

**Comment by groovecoder** [comment]
When I clicked the "Pay me!" button, I got ...
```
Internal Server Error: /payout-request
Traceback (most recent call last):
  File "/Users/lcrouch/python/codesy/lib/python2.7/site-packages/django/core/handlers/base.py", line 149, in get_response
    response = self.process_exception_by_middleware(e, request)
  File "/Users/lcrouch/python/codesy/lib/python2.7/site-packages/django/core/handlers/base.py", line 147, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/Users/lcrouch/python/codesy/lib/python2.7/site-packages/django/views/decorators/csrf.py", line 58, in wrapped_view
    return view_func(*args, **kwargs)
  File "/Users/lcrouch/python/codesy/lib/python2.7/site-packages/django/views/generic/base.py", line 68, in view
    return self.dispatch(request, *args, **kwargs)
  File "/Users/lcrouch/python/codesy/lib/python2.7/site-packages/rest_framework/views.py", line 466, in dispatch
    response = self.handle_exception(exc)
  File "/Users/lcrouch/python/codesy/lib/python2.7/site-packages/rest_framework/views.py", line 463, in dispatch
    response = handler(request, *args, **kwargs)
  File "/Users/lcrouch/code/codesy/codesy/api/views.py", line 196, in post
    codesy_payout = Payout.objects.filter(claim=claim)[0]
  File "/Users/lcrouch/python/codesy/lib/python2.7/site-packages/django/db/models/query.py", line 297, in __getitem__
    return list(qs)[0]
IndexError: list index out of range
```

**Comment by groovecoder** [comment]
I promise I plan to get to this ... busy birthday weekend.

File: api/urls.py [hunk 5]
```diff
@@ -15,7 +15,7 @@
 
 custom_endpoint_url_patterns = patterns(
     '',
-    url(r'^bid-for-url/', views.BidAPIView.as_view()),
+    url(r'^bid-status/', views.BidAPIView.as_view()),
```
Comment (jdungan) [inline api/urls.py]: This requires codesy/widget#91

**Comment by groovecoder** [comment]
Offer says "You paid 1.00 ... Fees: Stripe: 0.34 codesy: 0.02"

It should say "You paid 1.36"

**Comment by groovecoder** [comment]
OfferFee could use a better `__unicode__`

**Comment by groovecoder** [comment]
When I made an offer, the Stripe call failed, but codesy recorded it as success.

File: auctions/models.py [hunk 6]
```diff
@@ -165,6 +243,71 @@ def expires(self):
...
+        )
+        codesy_fee.save()
+
+        total_fees = paypal_fee.amount + codesy_fee.amount
+        codesy_payout.amount = codesy_payout.amount - total_fees
+        codesy_payout.save()
+        # attempt paypay payout
+        # paypal id are limited to 30 chars
+        # TODO: Fix this potential non-unique key
+        paypay_key = str(codesy_payout.transaction_key)[:30]
```
Comment (groovecoder) [inline auctions/models.py]: `paypay_key` should be `paypal_key`

**Comment by groovecoder** [comment]
When there was a PayPal API error during payout, there was no UI error.

**Comment by groovecoder** [comment]
Need to implement api_success for Stripe transactions.

**Comment by groovecoder** [comment]
Payout.trasnaction_key is storing the full key value, while PayPal only keeps the first [30:] characters.

File: auctions/tests/model_tests.py [hunk 7]
```diff
@@ -399,3 +409,72 @@ def test_no_votes_not_counted(self, mock_send_mail):
...
+        self.issue = mommy.make(Issue, url=self.url)
+        self.bid = mommy.make(
+            Bid, user=self.user1,
+            url=self.url, issue=self.issue, offer=50)
+
+    def test_key_created(self):
+        offer = mommy.make(Offer, bid=self.bid)
+        self.assertIsNotNone(offer.transaction_key)
+
+    @fudge.patch('stripe.Charge.create')
```
Comment (groovecoder) [inline auctions/tests/model_tests.py]: Needs to be `auctions.models.stripe.Charge.create`

Comment (groovecoder) [inline auctions/tests/model_tests.py]: This should fix https://github.com/codesy/codesy/issues/279

File: auctions/tests/model_tests.py [hunk 8]
```diff
@@ -399,3 +409,72 @@ def test_no_votes_not_counted(self, mock_send_mail):
...
+    def test_bid_fees(self, mock_stripe_create):
```
Comment (groovecoder) [inline auctions/tests/model_tests.py]: These tests depend on the django signal receiver framework. A more ideal way to test would be:
```
def test_bid_fees(self, mock_stripe_create):
    get_payment_for_offer(Bid, self.bid)
    # the rest of the assertions
```

File: auctions/tests/model_tests.py [hunk 9]
```diff
@@ -399,3 +409,72 @@ def test_no_votes_not_counted(self, mock_send_mail):
...
+        offer = Offer.objects.get(bid=self.bid)
+        self.assertEqual(offer.amount, 50)
+        fees = OfferFee.objects.filter(offer=offer)
+        self.assertEqual(len(fees), 2)
+        sum_fees = fees.aggregate(Sum('amount'))['amount__sum']
+        offer_with_fees = offer.amount + sum_fees
+        self.assertEqual(offer_with_fees, offer.charge_amount)
+
+    @fudge.patch('stripe.Charge.create')
+    def test_new_offer(self, mock_stripe_create):
```
Comment (groovecoder) [inline auctions/tests/model_tests.py]: could be `test_updated_offer_on_bid_creates_new_offer`

File: auctions/tests/model_tests.py [hunk 10]
```diff
@@ -399,3 +409,72 @@ def test_no_votes_not_counted(self, mock_send_mail):
...
+
+        self.bid1 = mommy.make(
+            Bid, user=self.user1,
+            url=self.url, issue=self.issue, ask=50)
+
+        self.bid2 = mommy.make(
+            Bid, user=self.user2,
+            url=self.url, issue=self.issue, offer=50)
+
+    @fudge.patch('paypalrestsdk.Payout.create')
```
Comment (groovecoder) [inline auctions/tests/model_tests.py]: `@fudge.patch('auctions.models.paypalrestsdk.PaypalPayout')` ?

Then ...

`mock_paypal.expects_call().returns(fudge.Fake().expects('create').returns(True))`

File: auctions/tests/model_tests.py [hunk 11]
```diff
@@ -399,3 +409,72 @@ def test_no_votes_not_counted(self, mock_send_mail):
...
+        mock_paypal.is_callable()
+        claim = mommy.make(Claim, user=self.user1, issue=self.issue)
+        claim.request_payout()
+        payouts = Payout.objects.filter(claim=claim)
+        self.assertEqual(len(payouts), 1)
+        payout = payouts[0]
+        fees = PayoutFee.objects.filter(payout=payout)
+        self.assertEqual(len(fees), 2)
+        sum_fees = fees.aggregate(Sum('amount'))['amount__sum']
+        self.assertEqual(sum_fees + payout.amount, self.bid1.ask)
```
Comment (groovecoder) [inline auctions/tests/model_tests.py]: should add a `def test_request_payout_failure`

File: codesy/base/models.py [hunk 12]
```diff
@@ -1,26 +1,43 @@
...
-    stripe_card_token = models.CharField(max_length=100, blank=True)
-    # TODO: remove/replace User.balanced_bank_account_href
-    balanced_bank_account_href = models.CharField(max_length=100, blank=True)
-
+    stripe_account_token = models.CharField(max_length=100, blank=True)
     USERNAME_FIELD = 'username'
 
 
+@receiver(pre_save, sender=settings.AUTH_USER_MODEL)
+def replace_cc_token_with_account_token(sender, instance, **kwargs):
```
Comment (groovecoder) [inline codesy/base/models.py]: This should be moved to separate endpoints so we aren't using the :sparkles: of temporarily storing the one-time CC token in the `stripe_account_token` field.

File: api/views.py [hunk 13]
```diff
@@ -158,3 +164,22 @@ def get(self, request, format=None):
...
+        try:
+            claim = Claim.objects.get(id=request.POST['claim'])
+            if request.user == claim.user:
+                claim.request_payout()
+        except:
+            pass
+        # these do not work
+        # return reverse('claim-status', pk=claim.id)
+        # return reverse('claim-status', args=[claim.id])
+        # return reverse('claim-status', kwargs={pk=claim.id})
```
Comment (groovecoder) [inline api/views.py]: `return redirect(reverse('claim-status', pk=claim.id))`

Comment (jdungan) [inline api/views.py]: 5c64e39

File: auctions/models.py (hunk 6 above)
Comment (jdungan) [inline auctions/models.py]: 1b08167

**Pull request merged**